Kernel/PCI/Access.cpp

```diff
@@ -99,43 +99,43 @@ u16 Access::early_read_type(Address address)
     return (early_read8_field(address, PCI_CLASS) << 8u) | early_read8_field(address, PCI_SUBCLASS);
 }
 
-void Access::enumerate_functions(int type, u8 bus, u8 slot, u8 function, Function<void(Address, ID)>& callback)
+void Access::enumerate_functions(int type, u8 bus, u8 device, u8 function, Function<void(Address, ID)>& callback, bool recursive)
 {
-    dbgln<PCI_DEBUG>("PCI: Enumerating function type={}, bus={}, slot={}, function={}", type, bus, slot, function);
-    Address address(0, bus, slot, function);
+    dbgln<PCI_DEBUG>("PCI: Enumerating function type={}, bus={}, device={}, function={}", type, bus, device, function);
+    Address address(0, bus, device, function);
     if (type == -1 || type == early_read_type(address))
         callback(address, { early_read16_field(address, PCI_VENDOR_ID), early_read16_field(address, PCI_DEVICE_ID) });
-    if (early_read_type(address) == PCI_TYPE_BRIDGE) {
+    if (early_read_type(address) == PCI_TYPE_BRIDGE && recursive) {
```

Comment:
Just wondering, if we have root PCI host bridge that declares itself as multi-function, does it mean that if encounter another PCI bridge when we scan a bus, it must be a "function" in the root PCI host bridge?

Reply:
From what I understand, the PCI bridge itself is a function of the root PCI host bridge, but its children are on a separate PCI bus. That matches what I've seen so far with my bare-metal target, but I have no idea just how flexible PCI can be in the wild.